PSICover/PSICover.cs

```diff
@@ -163,31 +165,112 @@ class Analyzer {
          var code = File.ReadAllLines (file);
          for (int i = 0; i < code.Length; i++)
             code[i] = code[i].Replace ('<', '\u00ab').Replace ('>', '\u00bb');
+         int hitCount = 0;
          foreach (var block in blocks) {
             bool hit = hits[block.Id] > 0;
-            string tag = $"<span class=\"{(hit ? "hit" : "unhit")}\">";
-            code[block.ELine] = code[block.ELine].Insert (block.ECol, "</span>");
-            code[block.SLine] = code[block.SLine].Insert (block.SCol, tag);
+            if (hit) hitCount++;
+            if (block.SLine == block.ELine) { // For same line blocks
+               string tag = $"<span class=\"{(hit ? "hit tooltip covered" : "unhit tooltip uncovered")}\" data-title=\"{(hit ? hits[block.Id] + " hits" : "Code block not hit")}\">";
+               code[block.ELine] = code[block.ELine].Insert (block.ECol, "</span>");
+               code[block.SLine] = code[block.SLine].Insert (block.SCol, tag);
+            } else { // for multi line blocks w/wo other blocks on same line
+               string blockTag = $"<span class=\"tooltip covered\" data-title=\"{(hit ? hits[block.Id] + " hits" : "Code block not hit")}\"><span class=\"{(hit ? "hit" : "unhit")}\">";
+               string insideBlockTag = $"<span class=\"{(hit ? "hit" : "unhit")}\">";
+               code[block.ELine] = code[block.ELine].Insert (block.ECol, "</span></span>");
+               code[block.ELine] = code[block.ELine].Insert (GetFirstNonWhitespaceIndex (code[block.ELine]), insideBlockTag);
+               int lastLine = block.ELine - 1;
+               while (lastLine > block.SLine) {
+                  if (!string.IsNullOrEmpty (code[lastLine])) {
+                     int sIndex = GetFirstNonWhitespaceIndex (code[lastLine]);
+                     if (sIndex != code[lastLine].Length) {
+                        int lIndex = GetLastNonWhitespaceIndex (code[lastLine]);
+                        code[lastLine] = code[lastLine].Insert (lIndex + 1, "</span>");
+                        code[lastLine] = code[lastLine].Insert (sIndex, insideBlockTag);
+                     }
+                  }
+                  lastLine--;
+               }
+               code[block.SLine] = code[block.SLine].Insert (GetLastNonWhitespaceIndex (code[block.SLine]) + 1, "</span>");
+               code[block.SLine] = code[block.SLine].Insert (block.SCol, blockTag);

```

Comment:
You are first printing last line and then middle lines and then fist lines. This can be done using for loop. Try to implement it with for loopl.